### PR TITLE
Global search

### LINKED ISSUE #2589: [Product Note] Omnibar Help Center
# Problem

New users who use appsmith have tons of questions about how to do x on the platform. Some of them reach out on discord but many of them simply leave. We want to make it easier for these users to have all their questions about the product answered.

# Solution

### **Build an FAQ knowledge base**

The knowledge base can be structured around 

- How to use a widget in various cases
    - Paginate a table
    - Filter a table
    - Upload a file to S3
    - Validate a form
    - Build dynamic forms
- How to use actions in various cases
    - Run APIs in parallel / serial
    - Chain multiple actions
    - Store action results
    - Send data from one page to another
    - Set Authentication for APIs
    - Passing params to `APIs` & accessing them using `this.params`
- How to troubleshoot errors in appsmith

The knowledge base content should have videos / gifs to better explain concepts

### Improve discovery of the knowledge base
Making use of the knowledge base a core part of the product building experience instead of an afterthought when you really have no choice. Developers should actively search for and use content to build their applications in the best way.
The documentation should be search friendly and the content searched for should be previewable and not only have the headings. The content should be keyboard accessible and easily dismissable. Searching, reading, and building should be a seamless flow for a user.

There needs to be a uniform way to discover content about a part of the application so that users don't have to guess where they can find this information. The content should be contextual to the part of the app and accessible in a predictable manner.
- API Pane
  - Pagination
  - Post Body
  - Settings
- Query Pane
  - Settings
- Property Pane
  - Properties
- Datasource Pane
  - Fields 

We can leverage the omnibar feature to achieve this.
https://www.figma.com/file/WYbFYKL9R8QanG15TMwalx/Omnibar-v1?node-id=0%3A1
We'll focus only on navigation and discovery of documentation in this version.

### Provide Cheatsheets
Not all questions require complex guides, some of them simply require an example for users to understand how it works. Cheat sheets as content will focus on providing short examples of commonly used functionality inside appsmith. Should the example not suffice, they will link to the complete guide

**Chain APIs**

    ```jsx
    API.run signature: (onSuccess: Function, onError: Function)

    {{ API1.run(() => API2.run(), () => showAlert("This was an error")) }}
    ```

**Send data to another page**
    ```jsx
    Navigate signature: (location: string, queryparams: Object)

    {{ navigate('Page1', { id: Table1.selectedRow.id }) }}
    ```

 **Access query parameters**

    ```jsx
    URL: https://app.appsmith.com/applications/xyz/pages/abc?id=36238

    {{ appsmith.URL.queryParams.id }}
    ```

**Other Cheats**
    - Store data in an app
    - access data from the store
    - Filtering data in a table
    - Formatting Table Data
    - Displaying data in a dropdown
    - Walk down a JSON structure - example {{api1.data.first.second}}

### PR DESCRIPTION
**Pull request by rishabhsaxena**
## Description
Search:
- Common search input to search documentation and entities (pages, actions, widgets, and data sources)

Default state:
- If the search input is empty, we show a list of default docs and recent entities.

Recent entities:
- An entity is added to the `recent list` on navigating to the entity page or selecting a widget 
- Recent entities are persisted on the client against the appId and removed when the app is deleted

Docs preview:
- Some custom logic apart from `markdown -> html`, done by `marked`, including :
  - Stripping description and gitbook plugins, and algolia highlight tags b/w pre/code sections 
  - Fix links 
  - Append document title and open documentation button

Hotkeys:
- `Up / down` arrow to change the active item
- `(Cmd / Ctrl) + K` keys to open the help center
- `Esc` close

Edge cases:
- By default navigating to a `datasource search item` would navigate to the `current selected page`
- A `datasource recent item` would be linked to a page
- On selecting an entity on another page, both the page as well the entity would be added to the recent list

Issues not picked with this PR:
- Navigating to a widget within another page doesn't scroll to it

Fixes #2589

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- spec added to test navigation to entities and show/hide omnibar modal

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/components/editorComponents/GlobalSearch/Description.tsx [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+      >
+        ✨ Press{" "}
+        <kbd
+          style={{
+            marginLeft: theme.spaces[1],
+            marginRight: theme.spaces[1],
+            color: "#3DA5D9",
+          }}
+        >
+          ↵
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/Description.tsx:243]: use html entity for return key here `&#9166;`

File: app/client/src/components/editorComponents/GlobalSearch/Description.tsx [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+      <StyledHitEnterMessageContainer
+        style={{ display: "flex", alignItems: "center" }}
+      >
+        ✨ Press{" "}
+        <kbd
+          style={{
+            marginLeft: theme.spaces[1],
+            marginRight: theme.spaces[1],
+            color: "#3DA5D9",
+          }}
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/Description.tsx:241]: Use styled components and get/define colors from default theme

File: app/client/src/components/editorComponents/GlobalSearch/Description.tsx [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+};
+
+const replaceHintTagsWithCode = (text: string) => {
+  let result = text.replace(/{% hint .*?%}/, "```");
+  result = result.replace(/{% endhint .*?%}/, "```");
+  result = marked(result);
+  return result;
+};
+
+const getDocumentationPreviewContent = (
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/Description.tsx:147]: Move markdown transformation to a different file and add unit tests. Keep this component simple

File: app/client/src/components/editorComponents/GlobalSearch/GlobalSearchHotKeys.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import React from "react";
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/GlobalSearchHotKeys.tsx:1]: This should be defined in Editor/index.ts Maybe we can define these(and existing) globally in some hotkeys file and then import

File: app/client/src/components/editorComponents/GlobalSearch/HelpBar.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  color: ${(props) => props.theme.colors.globalSearch.helpBarText};
+  background: ${(props) => props.theme.colors.globalSearch.helpBarBackground};
+  border: 1px solid ${(props) => props.theme.colors.globalSearch.helpBarBorder};
+  border-radius: ${(props) => props.theme.radii[1]}px;
+  height: 28px;
+  flex: 1;
+  max-width: 350px;
+`;
+
+const modText = () => (isMac() ? `⌘` : "ctrl");
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/HelpBar.tsx:29]: Use html entities for the command key

File: app/client/src/components/editorComponents/GlobalSearch/SearchBox.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    let timer: number;
+    if (modalOpen) {
+      timer = setTimeout(() => setListenToChange(true), 100);
+    }
+
+    return () => {
+      if (timer) clearTimeout(timer);
+      setListenToChange(false);
+    };
+  }, [modalOpen]);
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/GlobalSearch/SearchBox.tsx:60]: Compose as a single hook

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionRequest.java [hunk 7]
```diff
@@ -1,26 +1,19 @@
 package com.appsmith.external.models;
 
-import lombok.AllArgsConstructor;
+import com.fasterxml.jackson.databind.JsonNode;
```
Comment (hetunandu) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/ActionExecutionRequest.java:3]: I think these changes are here by mistake, can we pull the latest release?

File: app/client/src/sagas/InitSagas.ts [hunk 8]
```diff
@@ -218,12 +225,18 @@ export function* initializeAppViewerSaga(
...
+}
+
 export default function* watchInitSagas() {
   yield all([
     takeLatest(ReduxActionTypes.INITIALIZE_EDITOR, initializeEditorSaga),
     takeLatest(
       ReduxActionTypes.INITIALIZE_PAGE_VIEWER,
       initializeAppViewerSaga,
     ),
+    takeLatest(ReduxActionTypes.RESET_EDITOR_REQUEST, resetEditorSaga),
```
Comment (hetunandu) [inline app/client/src/sagas/InitSagas.ts:240]: Why is this needed? I could not find where this is called from

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/src/sagas/InitSagas.ts (hunk 8 above)
Comment (rishabhsaxena) [reply app/client/src/sagas/InitSagas.ts:240]: This is to reset `is editor initialized` and `are entities restored` while switching applications, call it from `componentWillMount` within the Editor

File: app/client/src/components/editorComponents/GlobalSearch/GlobalSearchHotKeys.tsx (hunk 4 above)
Comment (hetunandu) [reply app/client/src/components/editorComponents/GlobalSearch/GlobalSearchHotKeys.tsx:1]: This was not applied

**Review by hetunandu — COMMENTED** [review]
LGTM

**Review by hetunandu — APPROVED** [review]

**Pull request merged**